File: kotlin/client/src/main/kotlin/convex/kotlin/ConvexClient.kt
package convex.kotlin

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.put
import java.net.URI
import java.net.http.HttpClient
import java.net.http.HttpRequest
import java.net.http.HttpResponse
import java.time.Duration
import java.util.concurrent.atomic.AtomicBoolean

private const val MAX_RESPONSE_BYTES = 2 * 1024 * 1024

/**
 * An educational native Kotlin Convex client. It implements the documented JSON
 * HTTP endpoints itself and hands the experimental Live profile to [LiveManager].
 */
class ConvexClient(
    deploymentUrl: String,
    private val clientVersion: String = "kotlin-0.1.0",
    private val http: HttpClient = HttpClient.newBuilder().connectTimeout(Duration.ofSeconds(10)).build(),
) : AutoCloseable {
    internal val deploymentUrl = normaliseDeploymentUrl(deploymentUrl)
    private val closed = AtomicBoolean(false)

    @Volatile private var authToken = ""

    @Volatile private var live: LiveManager? = null

    fun setAuth(token: String) {
        if (closed.get()) throw ClosedException()
        authToken = token
    }

    fun query(
        path: String,
        args: JsonObject = JsonObject(emptyMap()),
    ) = call("query", path, args)

    fun mutation(
        path: String,
        args: JsonObject = JsonObject(emptyMap()),
    ) = call("mutation", path, args)

    fun action(
        path: String,
        args: JsonObject = JsonObject(emptyMap()),
    ) = call("action", path, args)

    private fun call(
        operation: String,
        path: String,
        args: JsonObject,
    ): Result {
        require(path.isNotBlank()) { "Convex function path is required" }
        if (closed.get()) throw ClosedException()
        val body =
            json.encodeToString(
                JsonObject.serializer(),
                buildJsonObject {
                    put("path", path)
                    put("args", args)
                    put("format", "json")
                },
            )
        val request =
            HttpRequest
                .newBuilder(URI("$deploymentUrl/api/$operation"))
                .timeout(Duration.ofSeconds(30))
                .header("Content-Type", "application/json")
                .header("Accept", "application/json")
                .header("Convex-Client", clientVersion)
                .apply { if (authToken.isNotBlank()) header("Authorization", "Bearer $authToken") }
                .POST(HttpRequest.BodyPublishers.ofString(body))
                .build()
        val response =
            try {
                http.send(request, HttpResponse.BodyHandlers.ofByteArray())
            } catch (error: Exception) {
                throw TransportException(operation, error)
            }
        if (response.body().size > MAX_RESPONSE_BYTES) {
            throw TransportException(operation, IllegalStateException("response exceeds $MAX_RESPONSE_BYTES bytes"))
        }
        val decoded =
            try {
                json.parseToJsonElement(response.body().decodeToString()).jsonObject
            } catch (error: Exception) {
                throw TransportException(operation, IllegalStateException("HTTP ${response.statusCode()} returned non-Convex JSON", error))
            }
        val logs = decoded["logLines"]?.let { value -> value.jsonArray.map { it.jsonPrimitive.content } } ?: emptyList()
        return when (decoded["status"]?.jsonPrimitive?.content) {
            "success" -> Result(decoded["value"] ?: throw ProtocolException("success response omitted value"), logs)
            "error" -> throw FunctionException(
                operation,
                decoded["errorMessage"]?.jsonPrimitive?.content ?: "Convex function failed",
                decoded["errorData"],
                logs,
            )
            else -> throw ProtocolException("HTTP ${response.statusCode()} response has an unknown status")
        }
    }

    fun subscribe(
        path: String,
        args: JsonObject = JsonObject(emptyMap()),
    ): Subscription {
        require(path.isNotBlank()) { "Convex function path is required" }
        if (closed.get()) throw ClosedException()
        return synchronized(this) {
            val manager = live ?: LiveManager(deploymentUrl, clientVersion).also { live = it }
            manager.subscribe(path, args)
        }
    }

    /** Adapter-only test hook; normal client builds do not expose an adapter executable. */
    internal fun debugDisconnectForAdapter() {
        if (closed.get()) throw ClosedException()
        (live ?: throw IllegalStateException("Live WebSocket has not been started")).debugDisconnect()
    }

    override fun close() {
        if (closed.compareAndSet(false, true)) {
            live?.close()
            http.shutdownNow()
            http.close()
        }
    }

    private fun normaliseDeploymentUrl(raw: String): String {
        val uri = URI(raw)
        require(uri.scheme in setOf("http", "https") && uri.host != null && uri.userInfo == null) {
            "Convex deployment URL must be an absolute HTTP(S) URL without user information"
        }
        return URI(uri.scheme, null, uri.host, uri.port, uri.path?.trimEnd('/'), null, null).toString().trimEnd('/')
    }

    companion object {
        /** Shared strict-ish JSON codec used by the client and its test-only adapter. */
        val json =
            Json {
                ignoreUnknownKeys = true
                explicitNulls = false
            }
    }
}

File: visual-basic-dotnet/client/ConvexClient.vb
Imports System.Net.Http
Imports System.Net.Http.Json
Imports System.Text.Json.Nodes

Namespace ConvexVisualBasic
    ''' <summary>A small native Visual Basic .NET client for Convex JSON HTTP functions.</summary>
    Public NotInheritable Class ConvexClient
        Implements IDisposable

        Private ReadOnly baseUri As Uri
        Private ReadOnly http As New HttpClient With {.Timeout = TimeSpan.FromSeconds(30)}
        Private token As String = ""
        Private closed As Boolean

        Public Sub New(deployment As String)
            Dim candidate As Uri = Nothing
            If Not Uri.TryCreate(deployment.TrimEnd("/"c), UriKind.Absolute, candidate) OrElse
                (candidate.Scheme <> Uri.UriSchemeHttp AndAlso candidate.Scheme <> Uri.UriSchemeHttps) OrElse
                Not String.IsNullOrEmpty(candidate.UserInfo) Then
                Throw New ArgumentException("Convex deployment URL must be http(s), have a host, and omit user info")
            End If
            baseUri = candidate
        End Sub

        Public Sub SetAuth(value As String)
            EnsureOpen()
            token = If(value, "")
        End Sub

        Public Function Query(path As String, args As JsonObject) As Task(Of Result)
            Return CallFunction("query", path, args)
        End Function

        Public Function Mutation(path As String, args As JsonObject) As Task(Of Result)
            Return CallFunction("mutation", path, args)
        End Function

        Public Function Action(path As String, args As JsonObject) As Task(Of Result)
            Return CallFunction("action", path, args)
        End Function

        Private Async Function CallFunction(operation As String, path As String, args As JsonObject) As Task(Of Result)
            EnsureOpen()
            If String.IsNullOrWhiteSpace(path) Then Throw New ArgumentException("Convex function path is required")

            Using request As New HttpRequestMessage(HttpMethod.Post, New Uri(baseUri, "/api/" & operation))
                request.Content = JsonContent.Create(New JsonObject From {
                    {"path", path}, {"args", args.DeepClone()}, {"format", "json"}
                })
                request.Headers.TryAddWithoutValidation("Convex-Client", "visual-basic-dotnet-0.1.0")
                If token.Length <> 0 Then request.Headers.TryAddWithoutValidation("Authorization", "Bearer " & token)

                Dim response As HttpResponseMessage
                Try
                    response = Await http.SendAsync(request).ConfigureAwait(False)
                Catch ex As Exception
                    Throw New TransportException(operation, ex)
                End Try

                Using response
                    Dim decoded As JsonObject
                    Try
                        decoded = Await response.Content.ReadFromJsonAsync(Of JsonObject)().ConfigureAwait(False)
                        If decoded Is Nothing Then Throw New InvalidOperationException("empty response")
                    Catch ex As Exception
                        Throw New TransportException(operation, New InvalidOperationException("non-Convex HTTP response", ex))
                    End Try
                    Dim logs = DecodeLogs(decoded)
                    Dim status = decoded("status")?.GetValue(Of String)()
                    If status = "success" Then
                        If Not decoded.ContainsKey("value") Then Throw New ProtocolException("success response omitted value")
                        Return New Result(If(decoded("value") Is Nothing, Nothing, decoded("value").DeepClone()), logs)
                    End If
                    If status = "error" Then
                        Throw New FunctionException(operation, If(decoded("errorMessage")?.GetValue(Of String)(), "Convex function failed"), decoded("errorData"), logs)
                    End If
                    Throw New ProtocolException("HTTP " & CStr(CInt(response.StatusCode)) & " response has unknown status")
                End Using
            End Using
        End Function

        Friend Shared Function DecodeLogs(message As JsonObject) As IReadOnlyList(Of String)
            Dim values = message("logLines")?.AsArray()
            If values Is Nothing Then Return Array.Empty(Of String)()
            Return values.Select(Function(value) If(value Is Nothing, "", value.GetValue(Of String)())).ToArray()
        End Function

        Private Sub EnsureOpen()
            If closed Then Throw New ObjectDisposedException(NameOf(ConvexClient))
        End Sub

        Public Sub Dispose() Implements IDisposable.Dispose
            closed = True
            http.Dispose()
        End Sub

        Public NotInheritable Class Result
            Public Sub New(resultValue As JsonNode, resultLogs As IReadOnlyList(Of String))
                Value = resultValue
                Logs = resultLogs
            End Sub
            Public ReadOnly Property Value As JsonNode
            Public ReadOnly Property Logs As IReadOnlyList(Of String)
        End Class

        Public Class FunctionException
            Inherits Exception
            Public Sub New(operation As String, message As String, data As JsonNode, logs As IReadOnlyList(Of String))
                MyBase.New(message)
                Me.Operation = operation
                ErrorData = If(data Is Nothing, Nothing, data.DeepClone())
                Me.Logs = logs
            End Sub
            Public ReadOnly Property Operation As String
            Public ReadOnly Property ErrorData As JsonNode
            Public ReadOnly Property Logs As IReadOnlyList(Of String)
        End Class

        Public Class TransportException
            Inherits Exception
            Public Sub New(operation As String, inner As Exception)
                MyBase.New(inner.Message, inner)
                Me.Operation = operation
            End Sub
            Public ReadOnly Property Operation As String
        End Class

        Public Class ProtocolException
            Inherits Exception
            Public Sub New(message As String)
                MyBase.New(message)
            End Sub
        End Class
    End Class
End Namespace
